pymtl3/passes/sverilog/import_/ImportPass.py

```diff
@@ -0,0 +1,380 @@
+#=========================================================================
+# ImportPass.py
+#=========================================================================
+# Author : Peitian Pan
+# Date   : May 25, 2019
+"""Provide a pass that imports arbitrary SystemVerilog modules."""
+
+from __future__ import absolute_import, division, print_function
+
+import copy
+import importlib
+import linecache
+import os
+import shutil
+import subprocess
+import sys
+
+from pymtl3.passes.BasePass import BasePass
+from pymtl3.passes.rtlir import get_component_ifc_rtlir
+from pymtl3.passes.sverilog.errors import SVerilogImportError
+from pymtl3.passes.sverilog.util.utility import get_component_unique_name, make_indent
+
+from .helpers import (
+    gen_comb_input,
+    gen_comb_output,
+    gen_internal_line_trace_py,
+    gen_line_trace_py,
+    gen_packed_ports,
+    gen_signal_decl_c,
+    gen_signal_decl_py,
+    gen_signal_init_c,
+    gen_wire_decl_py,
+)
+
+try:
+  # Python 2
+  reload
+except NameError:
+  # Python 3
+  from importlib import reload
+
+
+class ImportPass( BasePass ):
+  """Import an arbitrary SystemVerilog module as a PyMTL component.
+
+  The import pass takes as input a PyMTL component hierarchy where
+  the components to be imported have an `import` entry in their parameter
+  dictionary( set by calling the `set_parameter` PyMTL API ).
+  This pass assumes the modules to be imported are located in the current
+  directory and have name {full_name}.sv where `full_name` is the name of
+  component's class concatanated with the list of arguments of its construct
+  method. It has the following format:
+      > {module_name}[__{parameter_name}_{parameter_value}]*
+  As an example, component mux created through `mux = Mux(Bits32, 2)` has a
+  full name `Mux__Type_Bits32__ninputs_2`.
+  The top module inside the target .sv file should also have a full name.
+  """
+  def __call__( s, top ):
+    s.top = top
+    if not top._dsl.constructed:
+      raise SVerilogImportError( top, 
+        "please elaborate design {} before applying the import pass!". \
+          format(top) )
+    ret = s.traverse_hierarchy( top )
+    if ret is None:
+      ret = top
+    return ret
+
+  def traverse_hierarchy( s, m ):
+    if hasattr(m, "_sverilog_import") and m._sverilog_import:
```

Comment:
Hmmm do we plan to remove the leading underscore? I guess we need to decide what flag we use ..

Reply:
We definitely don't want the leading underscore. And yes we need to decide on the flags...